### ISSUE TITLE
Understanding shared connection

### ISSUE THREAD

**Issue by BastienAr** [OptimalBits/bull#913]
_This is more a question than an issue_

## Context

When you share redis connections and close all `Queues`, it does not close the redis connections (obviously since it's **shared**). So in my understanding it's up to you to close the redis connections after closing all the queues (e.g. before closing the app).

## Question

In `default` case of [the doc](https://github.com/OptimalBits/bull/blob/master/PATTERNS.md#reusing-redis-connections) we return a `new Redis()` instance, but we don't keep any reference to quit it afterwards. Is that safe to write something like:

```javascript
var {REDIS_URL} = process.env

var Redis = require('ioredis')
var client = new Redis(REDIS_URL);
var subscriber = new Redis(REDIS_URL);
var other = new Redis(REDIS_URL); // Reference to be able to quit the app

var opts = {
  createClient: function (type) {
    switch (type) {
      case 'client':
        return client;
      case 'subscriber':
        return subscriber;
      default:
        return other;
    }
  }
}

var queueFoo = new Queue('foobar', opts);
var queueQux = new Queue('quxbaz', opts);
```

If not, should I get the redis ref from some Queues after closing them ?

**Comment by manast** [comment]
Sorry for late response. Currently the shared clean ups is a bit crappy, I have it as a standing issue to fix it. When you call queue.close it will actually close all the connections independently of the ownership. Ideally it should work as you assumed, that the shared connections are not closed at all. I am by the way open for proposals on how to fix this in a appropriate way.

**Comment by BastienAr** [comment]
Hum, are you sure of that ? I do not have the possibility to retest right now, but I remember being stuck with an open connection to redis until I keep a reference and quit the default con myself.

And looking at the code, indeed all connection are closed here:
https://github.com/OptimalBits/bull/blob/9d07049be249029144b07c2efe230cd2c57bd9d2/lib/queue.js#L529-L544

But the con is added only if you don't pass the createClient arg:
https://github.com/OptimalBits/bull/blob/9d07049be249029144b07c2efe230cd2c57bd9d2/lib/queue.js#L269-L271

I may have miss something tho.

**Comment by manast** [comment]
no, you are right, the last if-statement was added recently, I had forgot about it.

**Comment by BastienAr** [comment]
Ok, and regarding my first question, I suppose it's safe to not construct a new Redis instance for each default case ?

**Comment by manast** [comment]
you may need to save them and close them manually later on.

**Comment by manast** [comment]
Please close if this issue was resolved.
